File: NOVEMBER-2025/22-11-2025/Minimum Operations to Connect Hospitals.py
class Solution:
    def minConnect(self, V, edges):
        parent = list(range(V))
        rank = [0] * V

        def find(x):
            if parent[x] != x:
                parent[x] = find(parent[x])
            return parent[x]

        def unite(a, b):
            pa, pb = find(a), find(b)
            if pa == pb:
                return
            if rank[pa] < rank[pb]:
                parent[pa] = pb
            elif rank[pb] < rank[pa]:
                parent[pb] = pa
            else:
                parent[pb] = pa
                rank[pa] += 1

        redundant = 0

        for u, v in edges:
            if find(u) == find(v):
                redundant += 1
            else:
                unite(u, v)

        components = sum(1 for i in range(V) if find(i) == i)
        need = components - 1

        return need if redundant >= need else -1

File: NOVEMBER-2025/22-11-2025/Minimum Operations to Connect Hospitals.java
class Solution {
    int[] parent, rank;

    int find(int x) {
        if (parent[x] != x) parent[x] = find(parent[x]);
        return parent[x];
    }
    void unite(int a, int b) {
        int pa = find(a), pb = find(b);
        if (pa == pb) return;
        if (rank[pa] < rank[pb]) parent[pa] = pb;
        else if (rank[pb] < rank[pa]) parent[pb] = pa;
        else {
            parent[pb] = pa;
            rank[pa]++;
        }
    }
    public int minConnect(int V, int[][] edges) {
        parent = new int[V];
        rank = new int[V];
        for (int i = 0; i < V; i++) parent[i] = i;
        int redundant = 0;
        for (int[] e : edges) {
            int u = e[0], v = e[1];
            if (find(u) == find(v)) redundant++;
            else unite(u, v);
        }
        int components = 0;
        for (int i = 0; i < V; i++) {
            if (find(i) == i) components++;
        }
        int need = components - 1;
        if (redundant >= need) return need;
        return -1;
    }
}